### PR TITLE
Support dictionary data structure

### PR DESCRIPTION
**Pull request by mxinden**
According to the [Swagger specification]
(http://swagger.io/specification/#model-with-map-dictionary-properties-88)
a dict is defined by the following json:

```
{
  "type": "object",
  "additionalProperties": {
    "type": "string"
  }
}
```
This PR checks if the field `additionalProperties` is defined and
`properties` is undefined. If so, it interprets it as a dict, otherwise
as an object as usual. It will use the nested field `type` as the second
type of the Elm dictionary. The above example would result in a
dictionary `Dict String String` with a decoder `dict string`.

Related to issue #10, related to PR #12, and [Alertmanager
PR](https://github.com/prometheus/alertmanager/pull/698)

### CODE REVIEW

**Comment by mxinden** [comment]
@ahultgren Porting my fix to new structure as discussed in PR #12. I will test it more tomorrow. Works with a new integration test already. 

Let me know what you think. Hope there is not a 'rewrite everything 2' :-P

File: source/Swagger/Decode.elm [hunk 1]
```diff
@@ -37,15 +37,17 @@ decodeType : Decoder Type
 decodeType =
     lazy
         (\_ ->
-            decode (,)
+            decode (,,,)
                 |> optional "type" string ""
                 |> maybe "$ref" string
+                |> maybe "properties" value
+                |> maybe "additionalProperties" value
```
Comment (ahultgren) [inline source/Swagger/Decode.elm]: Since these are just used in the object decoder, I would prefer to defer this to decodeObject (in the same way decodeString is implemented). So something like:

```elm
decodeObject : Decoder Type
decodeObject =
    decode (,,)
        |> optional "required" (list string) []
        |> maybe "properties" (lazy (\_ -> keyValuePairs decodeType))
        |> maybe "additionalProperties" (lazy (\_ -> decodeType))
        |> map objectOrDict
```

File: source/Generate/Type.elm [hunk 2]
```diff
@@ -50,6 +50,9 @@ renderType definition =
             Array_ items ->
                 typeAliasDecl <| list <| renderPropertyType name "Item" <| getItemsType items
 
+            Dict_ typeName ->
+                typeAliasDecl <| dict <| renderPropertyType name "Item" typeName
```
Comment (ahultgren) [inline source/Generate/Type.elm]: Use `type_` or similar since it refers to a `Type`. `typeName` would shadow the imported function.

File: source/Swagger/Decode.elm [hunk 3]
```diff
@@ -68,7 +70,11 @@ decodeTypeByType ( type_, ref ) =
                     decodeArray
 
                 _ ->
-                    lazy (\_ -> decodeObject)
+                    case (properties, additionalProperties) of
+                        (Nothing, Just _) ->
+                            decodeDict
+                        _ ->
+                            lazy (\_ -> decodeObject)
```
Comment (ahultgren) [inline source/Swagger/Decode.elm]: And this would instead become:

```elm
objectOrDict : ( List String, Maybe (List ( String, Type )), Maybe Type ) -> Type
objectOrDict ( required, properties, additionalProperties ) =
    case ( properties, additionalProperties ) of
        ( Nothing, Just addProps ) ->
            Dict_ addProps

        ( Just props, _ ) ->
            Object_ <| Properties <| List.map (property required) props

        _ ->
            Object_ <| Properties []
```

File: tests/Integration/Decoder.elm [hunk 4]
```diff
@@ -144,4 +163,7 @@ all =
         , test "should decode Rules" <|
             always <|
                 Expect.equal (Ok expectedRules) (decodeString decodeRules rulesJson)
+        , test "should decode labels" <|
+            always <|
+                Expect.equal (Ok expectedLabels) (decodeString decodeLabels labelsJson)
```
Comment (ahultgren) [inline tests/Integration/Decoder.elm:210]: A test case for an empty dict and a more complex one (object or ref) would be nice.

File: source/Generate/Decoder.elm [hunk 5]
```diff
@@ -69,6 +72,10 @@ renderArrayBody : String -> Type -> String
 renderArrayBody name type_ =
     "list " ++ (renderPropertyDecoder name "Item" type_)
 
+renderDictBody : String -> Type -> String
+renderDictBody name type_ =
+    "dict " ++ (renderPropertyDecoder name "Item" type_)
```
Comment (ahultgren) [inline source/Generate/Decoder.elm]: I think another name than "Item", perhaps "Property", would reduce risk of confusion with arrays.

**Review by ahultgren — CHANGES_REQUESTED** [review]
Except for the comments it's how I would have done it :)

Did you find it simpler to implement in the rewritten version?

**Comment by mxinden** [comment]
@ahultgren Thanks for the review. I will look into it tomorrow. Was definitely easier the second time - good work!

File: tests/Integration/Decoder.elm (hunk 4 above)
Comment (mxinden) [reply tests/Integration/Decoder.elm:210]: Added 3 more tests: empty dict, dict with object, dict with ref. Good idea! I was missing the flattening in my initial implementation. Fixed now.

**Comment by mxinden** [comment]
@ahultgren Addressed all your comments. Let me know what you think.

**Review by ahultgren — APPROVED** [review]

**Comment by ahultgren** [comment]
Awesome! Looks good 👍
